### PR TITLE
Update README for Windows build to improve usability

### PR DESCRIPTION
**Pull request by jcwchen**
**Description**
Provide more details in the README:
1. Recommend users to use Visual studio
2. Set the cmake flags 

Besides, since static protobuf of `3.12.x` work well with current onnx, update the protobuf version from `3.9.x` to `3.12.x`.

**Motivation and Context**
https://github.com/onnx/onnx/issues/2962
A user bumped into an error while installing onnx on Windows according to the README.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,15 +94,19 @@ pip install onnx
...
 
+Note that the instructions in this README assume you are using Visual Studio. It is recommended that you run all the commands from a shell started from "Developer Command Prompt for VS 2019" and keep the build system generator for cmake consistent.
+
 #### Build Protobuf and ONNX on Windows
 Step 1: Build Protobuf locally
 ```
 git clone https://github.com/protocolbuffers/protobuf.git
 cd protobuf
-git checkout 3.9.x
+git checkout 3.12.x
```
Comment (jcwchen) [inline README.md:104]: I can build onnx on Windows with the latest protobuf so I updated it. Please comment if you have any concern about this. Thanks.

File: README.md [hunk 2]
```diff
@@ -94,15 +94,19 @@ pip install onnx
 ### Windows
 If you are building ONNX from source on Windows, it is recommended that you also build Protobuf locally as a static library. The version distributed with conda-forge is a DLL and this is a conflict as ONNX expects it to be a static library.
 
+Note that the instructions in this README assume you are using Visual Studio. It is recommended that you run all the commands from a shell started from "Developer Command Prompt for VS 2019" and keep the build system generator for cmake consistent.
```
Comment (vinitra) [inline README.md:97]:
```suggestion
The instructions in this README assume you are using Visual Studio to build ONNX on Windows. It is recommended that you run all the commands from a shell started from "Developer Command Prompt for VS 2019" and keep the build system generator for cmake consistent.
```

A little unclear what you mean by the "build system generator for cmake"

Comment (jcwchen) [reply README.md]: Good catch. I will make an example. Thanks!

File: README.md (hunk 1 above)
Comment (wschin) [reply README.md:104]: Is this working with Windows Anaconda?

**Comment by advancedwebdeveloper** [comment]
There is nothing for Clang, on Windows.

I tried to compile ONNX and caught [https://github.com/pybind/pybind11/pull/1883](https://github.com/pybind/pybind11/pull/1883) .

That's CMakeFiles/onnx_cpp2py_export.dir/onnx/cpp2py_export.cc.obj build target (for Ninja, in my case):

[1/2] Building CXX object CMakeFiles/onnx_cpp2py_export.dir/onnx/cpp2py_export.cc.obj
FAILED: CMakeFiles/onnx_cpp2py_export.dir/onnx/cpp2py_export.cc.obj
C:\PROGRA~1\LLVM\bin\CLANG_~1.EXE -DONNXIFI_ENABLE_EXT=ON -DONNX_ML=1 -DONNX_NAMESPACE=onnx -DONNX_USE_LITE_PROTO=1 -Donnx_cpp2py_export_EXPORTS -I"C:\Program Files\Python38-32\include" -IC:\Users\clang\build_onnx\pybind11\include -IC:/Users/Public/Downloads/onnx-master -I. -isystem protoc/include -Wnon-virtual-dtor -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -fvisibility=hidden -std=c++1y -MD -MT CMakeFiles/onnx_cpp2py_export.dir/onnx/cpp2py_export.cc.obj -MF CMakeFiles\onnx_cpp2py_export.dir\onnx\cpp2py_export.cc.obj.d -o CMakeFiles/onnx_cpp2py_export.dir/onnx/cpp2py_export.cc.obj -c C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/checker.h:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/schema.h:24:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/shape_inference.h:4:
C:/Users/Public/Downloads/onnx-master\onnx/proto_utils.h:34:16: warning: 'SetTotalBytesLimit' is deprecated: Please use the single parameter version of SetTotalBytesLimit(). The second parameter is ignored. [-Wdeprecated-declarations]
  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
               ^
C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:69:9: note: in instantiation of function template specialization 'onnx::ParseProtoFromPyBytes<onnx::NodeProto>' requested here
        ParseProtoFromPyBytes(&proto, bytes);
        ^
protoc/include\google/protobuf/io/coded_stream.h:394:3: note: 'SetTotalBytesLimit' has been explicitly marked deprecated here
  PROTOBUF_DEPRECATED_MSG(
  ^
protoc/include\google/protobuf/port_def.inc:156:53: note: expanded from macro 'PROTOBUF_DEPRECATED_MSG'
#define PROTOBUF_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
                                                    ^
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/checker.h:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/schema.h:24:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/shape_inference.h:4:
C:/Users/Public/Downloads/onnx-master\onnx/proto_utils.h:34:16: warning: 'SetTotalBytesLimit' is deprecated: Please use the single parameter version of SetTotalBytesLimit(). The second parameter is ignored. [-Wdeprecated-declarations]
  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
               ^
C:/Users/Public/Downloads/onnx-master\onnx/py_utils.h:16:10: note: in instantiation of function template specialization 'onnx::ParseProtoFromBytes<onnx::NodeProto>' requested here
  return ParseProtoFromBytes(proto, buffer, length);
         ^
C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:69:9: note: in instantiation of function template specialization 'onnx::ParseProtoFromPyBytes<onnx::NodeProto>' requested here
        ParseProtoFromPyBytes(&proto, bytes);
        ^
protoc/include\google/protobuf/io/coded_stream.h:394:3: note: 'SetTotalBytesLimit' has been explicitly marked deprecated here
  PROTOBUF_DEPRECATED_MSG(
  ^
protoc/include\google/protobuf/port_def.inc:156:53: note: expanded from macro 'PROTOBUF_DEPRECATED_MSG'
#define PROTOBUF_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
                                                    ^
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/checker.h:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/schema.h:24:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/shape_inference.h:4:
C:/Users/Public/Downloads/onnx-master\onnx/proto_utils.h:34:16: warning: 'SetTotalBytesLimit' is deprecated: Please use the single parameter version of SetTotalBytesLimit(). The second parameter is ignored. [-Wdeprecated-declarations]
  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
               ^
C:/Users/Public/Downloads/onnx-master\onnx/py_utils.h:16:10: note: in instantiation of function template specialization 'onnx::ParseProtoFromBytes<onnx::ValueInfoProto>' requested here
  return ParseProtoFromBytes(proto, buffer, length);
         ^
C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:216:9: note: in instantiation of function template specialization 'onnx::ParseProtoFromPyBytes<onnx::ValueInfoProto>' requested here
        ParseProtoFromPyBytes(&proto, bytes);
        ^
protoc/include\google/protobuf/io/coded_stream.h:394:3: note: 'SetTotalBytesLimit' has been explicitly marked deprecated here
  PROTOBUF_DEPRECATED_MSG(
  ^
protoc/include\google/protobuf/port_def.inc:156:53: note: expanded from macro 'PROTOBUF_DEPRECATED_MSG'
#define PROTOBUF_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
                                                    ^
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/checker.h:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/schema.h:24:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/shape_inference.h:4:
C:/Users/Public/Downloads/onnx-master\onnx/proto_utils.h:34:16: warning: 'SetTotalBytesLimit' is deprecated: Please use the single parameter version of SetTotalBytesLimit(). The second parameter is ignored. [-Wdeprecated-declarations]
  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
               ^
C:/Users/Public/Downloads/onnx-master\onnx/py_utils.h:16:10: note: in instantiation of function template specialization 'onnx::ParseProtoFromBytes<onnx::TensorProto>' requested here
  return ParseProtoFromBytes(proto, buffer, length);
         ^
C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:224:9: note: in instantiation of function template specialization 'onnx::ParseProtoFromPyBytes<onnx::TensorProto>' requested here
        ParseProtoFromPyBytes(&proto, bytes);
        ^
protoc/include\google/protobuf/io/coded_stream.h:394:3: note: 'SetTotalBytesLimit' has been explicitly marked deprecated here
  PROTOBUF_DEPRECATED_MSG(
  ^
protoc/include\google/protobuf/port_def.inc:156:53: note: expanded from macro 'PROTOBUF_DEPRECATED_MSG'
#define PROTOBUF_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
                                                    ^
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/checker.h:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/schema.h:24:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/shape_inference.h:4:
C:/Users/Public/Downloads/onnx-master\onnx/proto_utils.h:34:16: warning: 'SetTotalBytesLimit' is deprecated: Please use the single parameter version of SetTotalBytesLimit(). The second parameter is ignored. [-Wdeprecated-declarations]
  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
               ^
C:/Users/Public/Downloads/onnx-master\onnx/py_utils.h:16:10: note: in instantiation of function template specialization 'onnx::ParseProtoFromBytes<onnx::SparseTensorProto>' requested here
  return ParseProtoFromBytes(proto, buffer, length);
         ^
C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:232:9: note: in instantiation of function template specialization 'onnx::ParseProtoFromPyBytes<onnx::SparseTensorProto>' requested here
        ParseProtoFromPyBytes(&proto, bytes);
        ^
protoc/include\google/protobuf/io/coded_stream.h:394:3: note: 'SetTotalBytesLimit' has been explicitly marked deprecated here
  PROTOBUF_DEPRECATED_MSG(
  ^
protoc/include\google/protobuf/port_def.inc:156:53: note: expanded from macro 'PROTOBUF_DEPRECATED_MSG'
#define PROTOBUF_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
                                                    ^
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/checker.h:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/schema.h:24:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/shape_inference.h:4:
C:/Users/Public/Downloads/onnx-master\onnx/proto_utils.h:34:16: warning: 'SetTotalBytesLimit' is deprecated: Please use the single parameter version of SetTotalBytesLimit(). The second parameter is ignored. [-Wdeprecated-declarations]
  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
               ^
C:/Users/Public/Downloads/onnx-master\onnx/py_utils.h:16:10: note: in instantiation of function template specialization 'onnx::ParseProtoFromBytes<onnx::AttributeProto>' requested here
  return ParseProtoFromBytes(proto, buffer, length);
         ^
C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:240:9: note: in instantiation of function template specialization 'onnx::ParseProtoFromPyBytes<onnx::AttributeProto>' requested here
        ParseProtoFromPyBytes(&proto, bytes);
        ^
protoc/include\google/protobuf/io/coded_stream.h:394:3: note: 'SetTotalBytesLimit' has been explicitly marked deprecated here
  PROTOBUF_DEPRECATED_MSG(
  ^
protoc/include\google/protobuf/port_def.inc:156:53: note: expanded from macro 'PROTOBUF_DEPRECATED_MSG'
#define PROTOBUF_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
                                                    ^
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/checker.h:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/schema.h:24:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/shape_inference.h:4:
C:/Users/Public/Downloads/onnx-master\onnx/proto_utils.h:34:16: warning: 'SetTotalBytesLimit' is deprecated: Please use the single parameter version of SetTotalBytesLimit(). The second parameter is ignored. [-Wdeprecated-declarations]
  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
               ^
C:/Users/Public/Downloads/onnx-master\onnx/py_utils.h:16:10: note: in instantiation of function template specialization 'onnx::ParseProtoFromBytes<onnx::GraphProto>' requested here
  return ParseProtoFromBytes(proto, buffer, length);
         ^
C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:257:9: note: in instantiation of function template specialization 'onnx::ParseProtoFromPyBytes<onnx::GraphProto>' requested here
        ParseProtoFromPyBytes(&proto, bytes);
        ^
protoc/include\google/protobuf/io/coded_stream.h:394:3: note: 'SetTotalBytesLimit' has been explicitly marked deprecated here
  PROTOBUF_DEPRECATED_MSG(
  ^
protoc/include\google/protobuf/port_def.inc:156:53: note: expanded from macro 'PROTOBUF_DEPRECATED_MSG'
#define PROTOBUF_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
                                                    ^
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/checker.h:7:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/schema.h:24:
In file included from C:/Users/Public/Downloads/onnx-master\onnx/defs/shape_inference.h:4:
C:/Users/Public/Downloads/onnx-master\onnx/proto_utils.h:34:16: warning: 'SetTotalBytesLimit' is deprecated: Please use the single parameter version of SetTotalBytesLimit(). The second parameter is ignored. [-Wdeprecated-declarations]
  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
               ^
C:/Users/Public/Downloads/onnx-master\onnx/py_utils.h:16:10: note: in instantiation of function template specialization 'onnx::ParseProtoFromBytes<onnx::ModelProto>' requested here
  return ParseProtoFromBytes(proto, buffer, length);
         ^
C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:264:5: note: in instantiation of function template specialization 'onnx::ParseProtoFromPyBytes<onnx::ModelProto>' requested here
    ParseProtoFromPyBytes(&proto, bytes);
    ^
protoc/include\google/protobuf/io/coded_stream.h:394:3: note: 'SetTotalBytesLimit' has been explicitly marked deprecated here
  PROTOBUF_DEPRECATED_MSG(
  ^
protoc/include\google/protobuf/port_def.inc:156:53: note: expanded from macro 'PROTOBUF_DEPRECATED_MSG'
#define PROTOBUF_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
                                                    ^
In file included from C:/Users/Public/Downloads/onnx-master/onnx/cpp2py_export.cc:1:
C:\Users\clang\build_onnx\pybind11\include\**pybind11/pybind11.h:1878:57: error: cannot compile this forwarded non-trivially copyable parameter yet**
    register_exception_translator([](std::exception_ptr p) {
                                                        ^

File: README.md (hunk 1 above)
Comment (jcwchen) [reply README.md:104]: cc @vinitra for checking with Windows Anaconda. Thanks!

**Comment by jcwchen** [comment]
@advancedwebdeveloper For building onnx on Windows, I suggest using `Visual Studio 16 19` as cmake generator because it is fully supported. Thanks.

File: README.md (hunk 1 above)
Comment (jcwchen) [reply README.md:104]: @wschin Even I just tested it. Even Protobuf 3.13 (shared libary) is working with Windows conda.

**Comment by jcwchen** [comment]
Two updates:
- Remove quotes for the environment setting to prevent failure on some Windows
- Mention `pip uinstall onnx` before building from source

Another TODO work item is to make sure the following setting is default:
`set CMAKE_ARGS=-DONNX_USE_PROTOBUF_SHARED_LIBS=OFF -DProtobuf_USE_STATIC_LIBS=ON`
Currently either build with shared protobuf or build with static protobuf needs to set `CMAKE_ARGS`. The default flags are somehow not working.

Thank @yuslepukhin for sharing the build experience.

File: README.md [hunk 3]
```diff
@@ -120,6 +126,9 @@ git submodule update --init --recursive
 set PATH=<protobuf_install_dir>\bin;%PATH%
 set USE_MSVC_STATIC_RUNTIME=0
 
+# use the static installed protobuf
+set CMAKE_ARGS=-DONNX_USE_PROTOBUF_SHARED_LIBS=OFF -DProtobuf_USE_STATIC_LIBS=ON
```
Comment (jcwchen) [inline README.md:130]: Defaults for `-DProtobuf_USE_STATIC_LIBS` and `-DONNX_USE_PROTOBUF_SHARED_LIBS` are not working... Thus the setting here is required.

File: README.md (hunk 1 above)
Comment (askhade) [reply README.md:106]: please use 3.11.3 (as this is the one being used in the current release package)

File: README.md [hunk 4]
```diff
@@ -94,15 +96,19 @@ pip install onnx
...
 Step 1: Build Protobuf locally
 ```
 git clone https://github.com/protocolbuffers/protobuf.git
 cd protobuf
-git checkout 3.9.x
+git checkout 3.12.x
 cd cmake
 # Explicitly set -Dprotobuf_MSVC_STATIC_RUNTIME=OFF to make sure protobuf does not statically link to runtime library
-cmake -G "Visual Studio 15 2017 Win64" -Dprotobuf_MSVC_STATIC_RUNTIME=OFF -Dprotobuf_BUILD_TESTS=OFF -Dprotobuf_BUILD_EXAMPLES=OFF -DCMAKE_INSTALL_PREFIX=<protobuf_install_dir>
+cmake -Dprotobuf_MSVC_STATIC_RUNTIME=OFF -Dprotobuf_BUILD_TESTS=OFF -Dprotobuf_BUILD_EXAMPLES=OFF -DCMAKE_INSTALL_PREFIX=<protobuf_install_dir>
```
Comment (askhade) [inline README.md:109]: if you are adding an example below then I suggest changing this to 
 cmake -G <generator> -A <arch> -Dprotobuf_MSVC_STATIC_RUNTIME=OFF -Dprotobuf_BUILD_TESTS=OFF -Dprotobuf_BUILD_EXAMPLES=OFF -DCMAKE_INSTALL_PREFIX=<protobuf_install_dir>

File: README.md [hunk 5]
```diff
@@ -120,6 +126,9 @@ git submodule update --init --recursive
 set PATH=<protobuf_install_dir>\bin;%PATH%
```
Comment (askhade) [inline README.md:126]: I think \include and \libs also need to be added to the path

Comment (jcwchen) [reply README.md:126]: Only including the \bin path which has `protoc.exe` works in my local Windows... Are \include and \libs needed for building ONNX with static protobuf in GitHub Action CI?

**Review by askhade — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (askhade) [reply README.md:126]: yes they are needed. Please add them

**Review by askhade — APPROVED** [review]

**Comment by askhade** [comment]
LGTM 
suggestion: we can add a script for building onnx which sets all env variables etc... something for later

**Review by gramalingam — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (jcwchen) [reply README.md]: Added. Thanks

**Pull request merged**